### PR TITLE
Drivers for RGB 16 bit parallel displays

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

This PR provides drivers for RGB parallel displays, including:

* st7701s display, as used in the Seeed Sensecap Indicator and other larger LCD displays.
* So-called RPI_DPI_RGB displays, which are 16 bit parallel with no SPI config required.
 
The drivers uses the ESP-IDF RGB-LCD support. PSRAM is required.

Tested so far on the Sensecap Indicator, Waveshare ESP32-S3 Touch 4.3,  the Makerfabs 4" display and the Makerfabs 4.3" 480x800 display.

The st7701s driver moves most of the board-specific configuration into the YAML file, e.g. it is possible in YAML to specify vendor-specific initialisation sequences. This avoids having to add model-specific support in the C++ code.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3546
## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
display:
  - platform: st7701s
    auto_clear_enabled: false
    update_interval: never
    spi_mode: MODE3
    color_order: RGB
    dimensions:
      width: 480
      height: 480
    invert_colors: true
    transform:
      mirror_x: true
      mirror_y: true
    cs_pin:
      pca9554: p_c_a
      number: 4
    reset_pin:
      pca9554: p_c_a
      number: 5
    de_pin: 18
    hsync_pin: 16
    vsync_pin: 17
    pclk_pin: 21
    init_sequence:
      - 1 # select canned init sequence number 1
# Custom sequences are an array, first byte is command, the rest are data.
      - [ 0xE0, 0x1F ]  # Set sunlight readable enhancement
    data_pins:
      red:
        - 4         #r1
        - 3         #r2
        - 2         #r3
        - 1         #r4
        - 0         #r5
      green:
        - 10        #g0
        - 9         #g1
        - 8         #g2
        - 7         #g3
        - 6         #g4
        - 5         #g5
      blue:
        - 15        #b1
        - 14        #b2
        - 13        #b3
        - 12        #b4
        - 11        #b5
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mreditor97** [comment]
I am liking the look of this, I did some work on trying to get the display driver working but ended up running out of time to get it working. I have got a ST7701 RGB display (Lilygo T-RGB Round Display) so I will be happy to test it once you have got it ready!

**Comment by mreditor97** [comment]
Do you have an example of the format required within ESP Home to change the `init_sequence`, so I can create a sequence specific for my device?

I have seen that you have made it possible to do, I spent a little time trying to figure out the format of it but everything I tried wasn't being accepted in the YAML interface.

Thanks for your great work so far!!

**Comment by clydebarrow** [comment]
I have updated the sample config here: https://gist.github.com/clydebarrow/ef89e9a93bd44771483b9144ae9042a1

with an example of adding custom init sequences.

**Comment by clydebarrow** [comment]
this PR includes changes from #6034 and should not be merged until after that.

**Comment by DannyN2222** [comment]
hi, so saw this pr and was quite pleased as ive been wanting to use larger screens with esphome however noticed that this PR doesnt contain a test for the rpi_dpi_rgb compoent thats also added in this PR
anyways ive got a screen thats compatable with that (a https://www.elecrow.com/esp32-display-7-inch-hmi-display-rgb-tft-lcd-touch-screen-support-lvgl.html) and got it mostly working however its not drawn correctly using the lambda "      it.fill(Color::WHITE);
      it.filled_rectangle(0, 0, 320, 65, blue);" the display correctly draws the white box and flashs the blue box, from using update inverval never, i can see what its doing its redrawing all the white background then drawing the blue box on top as a second operation this seems different to previous display components where it was one continus update with all the text / icons ect drawn as one pass, this can be solved for displays that like mine rarly update by just calling update.component and setting update interval to never but for those people that want to use these as touch screens where loading menus and stuff, i can see this being pretty major, ill attach a video of this for clarity

**Comment by DannyN2222** [comment]
https://github.com/esphome/esphome/assets/73726451/6546376f-1f65-4c7b-ba2d-d35b9265c19a

**Comment by DannyN2222** [comment]
^^^ 
above is Update_interval: never (with 1 manually triggered update) 
Code: 
```
   lambda: |-
      it.fill(Color::WHITE);
      it.filled_rectangle(0, 0, 320, 65, blue);
```
(github isnt showing any comments on those with videos for me so ill repeat it here)
Below is Update_interval: 3s

Edit: im being dumb, i had the text selected when i imported the video so turns out that means the video overwritten the comment

**Comment by DannyN2222** [comment]
Update_interval: 3s
Code:
lambda: |-
it.fill(Color::WHITE);
it.filled_rectangle(0, 0, 320, 65, blue);
https://github.com/esphome/esphome/assets/73726451/045ba372-6e0a-4c0c-97aa-dff155d10dd2

**Comment by clydebarrow** [comment]
Yeah, you will want to use LVGL with this rather than the native ESPHome drawing primitives, which are not ideal for large displays.

**Comment by DannyN2222** [comment]
oh one more thing to note is that if rotation: is set to anything other than 0° (or ommited) the display is just drawn blank (black)

**Comment by DannyN2222** [comment]
> Yeah, you will want to use LVGL with this rather than the native ESPHome drawing primitives, which are not ideal for large displays.

Make sence, seems like that should be made clear in the documentation for it, otherwise people will thinks its a bug

**Comment by tiimsvk** [comment]
Any updates?

**Comment by clydebarrow** [comment]
> Any updates?

Waiting for review.

**Comment by tiimsvk** [comment]
i would like to test it as a custom_component. It is possible?
I tried importing it like this:
```
external_components:
  - source:  github://pr#5872
    components: [ st7701s ]

display:
  - platform: st7701s
    .....
```


but the console says:
`  Platform not found: 'display.st7701s'.`
I also tried locally but no change.
Thanks


EDIT: add missing component "rpi_dpi_rgb" fix this problem

**Comment by tiimsvk** [comment]
I successfully tested the module from ali:
https://www.aliexpress.com/item/1005006302251538.html

Chip ST7701 Touch GT911 (not tested yet)
Connected pin display 40 pin header:
```
 # ----- DISPLAY -----
  # Description pins:
   # PIN NO | PIN NAME | DESCRIPTION     | GPIO         | DESCRIPTION        # PIN NO | PIN NAME | DESCRIPTION     | GPIO       | DESCRIPTION
   # 1      | VLED+    | LED ANODE       | KDG2V GPIO38 | BACKLIGHT DISPLAY  # 21     | DB5      | 18-BIT BUS      | GPIO15     |
   # 2      | VLED-    | LED CATHODE     | GND          |                    # 22     | DB6      | 18-BIT BUS      | GPIO08     |
   # 3      | VLED-    | LED CATHODE     | GND          |                    # 23     | DB7      | 18-BIT BUS      | GPIO20     |
   # 4      | GND      | GROUND          | GND          |                    # 24     | DB8      | 18-BIT BUS      | GPIO03     |
   # 5      | VCC      | POWER SUPPLY    | 3V3          |                    # 25     | DB9      | 18-BIT BUS      | GPIO46     |
   # 6      | RST      | RESET ACTIVE LOW| NC           |                    # 26     | DB10     | 18-BIT BUS      | GPIO09     |
   # 7      | NC       |                 | NC           |                    # 27     | DB11     | 18-BIT BUS      | GPIO10     |
   # 8      | NC       |                 | NC           |                    # 28     | DB12     | 18-BIT BUS      | NC         |
   # 9      | SDA      | DATA I/O BI SPI | GPIO47       |                    # 29     | DB13     | 18-BIT BUS      | GPIO11     |
   # 10     | SCK      | CLOCK I SPI     | GPIO48       |                    # 30     | DB14     | 18-BIT BUS      | GPIO12     |
   # 11     | CS       | CHIP SELECTED   | GPIO39       |                    # 31     | DB15     | 18-BIT BUS      | GPIO13     |
   # 12     | PCLK     | DOT CLOCK RGB   | GPIO21       |                    # 32     | DB16     | 18-BIT BUS      | GPIO14     |
   # 13     | DE       | DATA ENABLE     | GPIO18       |                    # 33     | DB17     | 18-BIT BUS      | GPIO0      | 
   # 14     | VSYNC    | FRAME SYNC      | GPIO17       |                    # 34     | GND      | GROUND          | GND        |
   # 15     | HSYNC    | LINE SYNC       | GPIO16       |                    # 35     | TP-INT   | TOUCH INTERRUPT | GND        |
   # 16     | DB0      | 18-BIT BUS      | NC           |                    # 36     | TP-SDA   | TOUCH SDA I2C   | GPIO19     | TOUCH SDA   GT911
   # 17     | DB1      | 18-BIT BUS      | GPIO04       |                    # 37     | TP-SCL   | TOUCH SCL I2C   | GPIO45     | TOUCH SCL   GT911
   # 18     | DB2      | 18-BIT BUS      | GPIO05       |                    # 38     | TP-RST   | TOUCH RESET     | RST        | TOUCH RESET GT911
   # 19     | DB3      | 18-BIT BUS      | GPIO06       |                    # 39     | TP-VCL   | POWER SUPPLY    | VDD        | TOUCH POWER SUPPLY
   # 20     | DB4      | 18-BIT BUS      | GPIO07       |                    # 40     | GND      | GROUND          | GND        |

# RGB DATA SIGNAL
 # DB0:  BLUE  LSB
 # DB5:  BLUE  MSB
 # DB6:  GREEN LSB
 # DB11: GREEN MSB
 # DB12: RED   LSB
 # DB17: RED   MSB
```

Questions:
- rendering is only possible in this slow way as mentioned in the post above? (I also changed the speed data_rate 40MHz)
- Why is the entire content rendered anew (when only a change (sensor number, etc.) would suffice during update_interval)
- The default background for rendering is white, is it possible to change it to black or any color? (fixed invert_color: false)
- I can't achieve the RGB scheme, am I making a mistake somewhere in the code? See Image


the small rectangles in the upper left are blue, red and green.

(I tried: 
```
display:
  invert_colors: true/false; 
  color_order: RGB/BGR
image:
  type: RGB24/RGB565
```

My current code::
```
esphome:
  name: assistant-voice
  friendly_name: assistant-voice
  platformio_options:
    board_build.flash_mode: dio

esp32:
  board: esp32-s3-devkitc-1
  variant: esp32s3
  flash_size: 16MB
  framework:
    type: esp-idf
    sdkconfig_options:
      CONFIG_ESP32S3_DEFAULT_CPU_FREQ_240: "y"
      CONFIG_ESP32S3_DATA_CACHE_64KB: "y"
      CONFIG_ESP32S3_DATA_CACHE_LINE_64B: "y"
      CONFIG_SPIRAM_FETCH_INSTRUCTIONS: y
      CONFIG_SPIRAM_RODATA: y

psram:
  mode: octal
  speed: 80MHz

logger:

api:

ota:

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

spi:
  - id: lcd_spi
    clk_pin: 48
    mosi_pin: 47

#-------------------------------------------
# EXTERNAL AND CUSTOM COMPONENT
#-------------------------------------------
external_components:
  - source:  github://pr#5872
    components: [ st7701s, rpi_dpi_rgb ]

output:
  - platform: ledc
    pin:
      number: GPIO38
    id: backlight
    frequency: 100Hz

light:
  - platform: monochromatic
    output: backlight
    name: Display Backlight
    id: disp_bl
    restore_mode: ALWAYS_ON

i2c:
  sda: 19
  scl: 45  # strapping
  id: bus_a

# TOUCH GT911
# 1. GND
# 2. SDA
# 3. SCL
# 4. RST = 
# 5. INT = GND
# 6. VDD

# I/O expander. Touchscreen RESET and INTERRUPT are not used.
pca9554:
  - id: p_c_a
    pin_count: 16
    address: 0x20

#touchscreen:
#  platform: gt911
#  id: my_touchscreen
#  interrupt_pin:
#    pca9554: p_c_a
#    number: 1 # CONNECTED TO GND

display:
  - platform: st7701s
    update_interval: 10s
    auto_clear_enabled: false
    data_rate: 40MHz
    spi_mode: MODE3
    color_order: RGB
    dimensions:
      width: 480
      height: 480
    invert_colors: false
    transform:
      mirror_x: true
      mirror_y: true
    cs_pin: 39
      # reset not defined 
    reset_pin: # CONNECTED TO GND
      pca9554: p_c_a
      number: 5
    de_pin: 18
    hsync_pin: 16
    vsync_pin: 17
    pclk_pin: 21
    init_sequence: 1
    data_pins:
      - 4
      - 5
      - 6
      - 7
      - 15
      - 8
      - 20
      - number: 3
        ignore_strapping_warning: true
      - number: 46 
        ignore_strapping_warning: true
      - 9
      - 10
      - 11
      - 12
      - 13
      - 14
      - number: 0 
        ignore_strapping_warning: true
    lambda: |-
      //--------------------------------------------------------------------------
      // START PAGE - log updating display
      //--------------------------------------------------------------------------
      ESP_LOGI("display", "Updating Display...");

      // Draw the image my_image at position [x=0,y=0]
      it.image(0, 0, id(esphome_logo));
      //it.fill(Color::RED);
      it.filled_rectangle(0, 0, 100, 50, blue);
      it.filled_rectangle(0, 50, 100, 50, red);
      it.filled_rectangle(0, 100, 100, 50, green);
      it.print(240, 240, id(my_font), id(time_color), TextAlign::CENTER,"Hello ESPHome!");


image:
  - file: image/back16.png
    id: esphome_logo
    resize: 480x480
    type: RGB24

font:
  - file: 'fonts/font_bd.ttf'
    id: my_font
    size: 54

color:
  - id: time_color
    red_int: 0
    green_int: 255
    blue_int: 0
    white_int: 0

  - id: blue
    red_int: 0
    green_int: 0
    blue_int: 255
    white_int: 0

  - id: red
    red_int: 255
    green_int: 0
    blue_int: 0
    white_int: 0

  - id: green
    red_int: 0
    green_int: 255
    blue_int: 0
    white_int: 0
```

Thanks for answer

**Comment by clydebarrow** [comment]
> * Why is the entire content rendered anew (when only a change (sensor number, etc.) would suffice during update_interval)

That's how the ESPHome display component works. You will need to use LVGL to get partial updating, which will be very fast. See https://discord.com/channels/429907082951524364/1119046373543780363

> 
> * The default background for rendering is white, is it possible to change it to black or any color? (fixed invert_color: false)

You will need to write your own background.

> 
> * I can't achieve the RGB scheme, am I making a mistake somewhere in the code? See Image

Probably you need to switch the upper and lower data bytes (move the first 8 bits below the second 8 bits.) The display is driven big-endian, See the docs: https://deploy-preview-3546--esphome.netlify.app/components/display/st7701s.html?highlight=st7701s

**Comment by clydebarrow** [comment]
The `data_pins:` schema has changed, now you just specify red, green and blue separately. See https://gist.github.com/clydebarrow/ef89e9a93bd44771483b9144ae9042a1#file-indicator-yaml for an example.

File: esphome/components/rpi_dpi_rgb/display.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+try:
+    from esphome.const import (
+        CONF_OFFSET_HEIGHT,
+        CONF_OFFSET_WIDTH,
+        CONF_INVERT_COLORS,
+    )
+except ImportError:
+    CONF_OFFSET_HEIGHT = "offset_height"
+    CONF_OFFSET_WIDTH = "offset_width"
+    CONF_INVERT_COLORS = "invert_colors"
```
Comment (jesserockz) [inline esphome/components/rpi_dpi_rgb/display.py]:
```suggestion
    CONF_NUMBER,
    CONF_OFFSET_HEIGHT,
    CONF_OFFSET_WIDTH,
    CONF_INVERT_COLORS,
```

File: esphome/components/rpi_dpi_rgb/display.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+COLOR_ORDERS = {
+    "RGB": ColorOrder.COLOR_ORDER_RGB,
+    "BGR": ColorOrder.COLOR_ORDER_BGR,
+}
+DATA_PIN_SCHEMA = pins.gpio_pin_schema(
+    {
+        CONF_OUTPUT: True,
+    },
+    internal=True,
+)
```
Comment (jesserockz) [inline esphome/components/rpi_dpi_rgb/display.py]: Why not use the built in types?
```suggestion
DATA_PIN_SCHEMA = pins.internal_gpio_output_pin_schema
```

File: esphome/components/st7701s/display.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+COLOR_ORDERS = {
+    "RGB": ColorOrder.COLOR_ORDER_RGB,
+    "BGR": ColorOrder.COLOR_ORDER_BGR,
+}
+DATA_PIN_SCHEMA = pins.gpio_pin_schema(
+    {
+        CONF_OUTPUT: True,
+    },
+    internal=True,
+)
```
Comment (jesserockz) [inline esphome/components/st7701s/display.py]:
```suggestion
DATA_PIN_SCHEMA = pins.internal_gpio_output_pin_schema
```

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**